test/e2e_node/device_plugin_test.go

```diff
@@ -589,6 +589,103 @@ func testDevicePlugin(f *framework.Framework, pluginSockDir string) {
 				framework.Fail(fmt.Sprintf("stale device assignment after pod deletion while kubelet was down allocated=%v error=%v", allocated, err))
 			}
 		})
+
+		ginkgo.It("Can schedule a pod with a restartable init container [NodeAlphaFeature:SidecarContainers]", func(ctx context.Context) {
+			podRECMD := "devs=$(ls /tmp/ | egrep '^Dev-[0-9]+$') && echo stub devices: $devs && sleep %s"
+			sleepOneSecond := "1s"
+			rl := v1.ResourceList{v1.ResourceName(SampleDeviceResourceName): *resource.NewQuantity(1, resource.DecimalSI)}
+			pod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{Name: "device-plugin-test-" + string(uuid.NewUUID())},
+				Spec: v1.PodSpec{
+					RestartPolicy: v1.RestartPolicyAlways,
+					InitContainers: []v1.Container{
+						{
+							Image:   busyboxImage,
+							Name:    "init-1",
+							Command: []string{"sh", "-c", fmt.Sprintf(podRECMD, sleepOneSecond)},
+							Resources: v1.ResourceRequirements{
+								Limits:   rl,
+								Requests: rl,
+							},
+						},
+						{
+							Image:   busyboxImage,
+							Name:    "restartable-init-2",
+							Command: []string{"sh", "-c", fmt.Sprintf(podRECMD, sleepIntervalForever)},
+							Resources: v1.ResourceRequirements{
+								Limits:   rl,
+								Requests: rl,
+							},
+							RestartPolicy: &containerRestartPolicyAlways,
+						},
+					},
+					Containers: []v1.Container{{
+						Image:   busyboxImage,
+						Name:    "regular-1",
+						Command: []string{"sh", "-c", fmt.Sprintf(podRECMD, sleepIntervalForever)},
+						Resources: v1.ResourceRequirements{
+							Limits:   rl,
+							Requests: rl,
+						},
+					}},
+				},
+			}
+
+			pod1 := e2epod.NewPodClient(f).CreateSync(ctx, pod)
+			deviceIDRE := "stub devices: (Dev-[0-9]+)"
+
+			devID1, err := parseLog(ctx, f, pod1.Name, pod1.Spec.InitContainers[0].Name, deviceIDRE)
+			framework.ExpectNoError(err, "getting logs for pod %q/%q", pod1.Name, pod1.Spec.InitContainers[0].Name)
+			gomega.Expect(devID1).To(gomega.Not(gomega.Equal("")), "pod1's init container requested a device but started successfully without")
+
+			devID2, err := parseLog(ctx, f, pod1.Name, pod1.Spec.InitContainers[1].Name, deviceIDRE)
+			framework.ExpectNoError(err, "getting logs for pod %q/%q", pod1.Name, pod1.Spec.InitContainers[1].Name)
+			gomega.Expect(devID2).To(gomega.Not(gomega.Equal("")), "pod1's restartable init container requested a device but started successfully without")
```

Comment:
Can you explain why we are okay with this? Is it because the device that was allocated to this container can be reused. I wonder if there is a way we can check if the init container was allocated a device at all?

Reply:
this expects the restartable init container to not start with empty devices.
In other words, the restartable init container should have some kind of devices allocated.

Reply:
Ah yeah, this is the error message in case devices are unexpected not allocated.